src/wpilib/joystick.rs

```
@@ -0,0 +1,238 @@
+        self.get_raw_button(7)
+    }
+
+    pub fn get_start_button(&mut self) -> bool {
+        self.get_raw_button(8)
+    }
+
+    pub fn get_axis_button(&mut self, side: JoystickSide) -> bool {
+        if side == JoystickSide::kLeftHand {
+            self.get_raw_button(9)
+        } else {
+            self.get_raw_button(10)
+        }
+    }
+
+    //axes
+    pub fn get_x(&mut self, side: JoystickSide) -> f32 {
+        if side == JoystickSide::kLeftHand {
+            self.get_raw_axis(0)
+        } else {
+            self.get_raw_axis(4)
+        }
+    }
+
+    pub fn get_y(&mut self, side: JoystickSide) -> f32 {
+        if side == JoystickSide::kLeftHand {
+            self.get_raw_axis(1)
+        } else {
+            self.get_raw_axis(5)
+        }
+    }
+
+    pub fn get_trigger(&mut self, side: JoystickSide) -> f32 {
+        if side == JoystickSide::kLeftHand {
+            self.get_raw_axis(2)
+        } else {
+            self.get_raw_axis(3)
+        }
+    }
+
+    pub fn get_dpad(&mut self) -> DPAD {
+        match self.get_pov(1) {
+            0 => DPAD::Up,
+            1 => DPAD::UpRight,
+            2 => DPAD::Right,
+            3 => DPAD::DownRight,
+            4 => DPAD::Down,
+            5 => DPAD::DownLeft,
+            6 => DPAD::Left,
+            7 => DPAD::UpLeft,
+            _ => DPAD::Neutral,
+        }
+    }
+ }
+
+impl JoystickBase for XBoxController {
+    fn get_raw_axis(&mut self, axis: usize) -> f32 {
+        match self.ds.get_joystick_axis(self.port, axis) {
+            Ok(val) => val,
+            Err(e) => 0f32,
+        }
+    }
+
+    fn get_raw_button(&mut self, button: usize) -> bool {
+        match self.ds.get_joystick_button(self.port, button) {
+            Ok(val) => val,
+            Err(e) => false,
+        }
+    }
+
+    fn get_pov(&mut self, pov: usize) -> i16 {
+        match self.ds.get_joystick_pov(self.port, pov) {
+            Ok(val) => val,
+            Err(e) => -01,
```

this looks suspiciously like a hardcoded error value. Why use this instead of an error type, or ignoring the error like the other `get_` methods do?

---

It is not a hard coded error value, rather, the idea was to return the value that corresponds with "neutral" on all the joystick get methods. As far as I'm aware, the default pov value is -01